### PR TITLE
Extracted chrome strings

### PR DESCRIPTION
**Pull request by Mailaender**
This process is not as automated as I would have wished for as this covers 4 mods including one that deviates the layout a lot, so there was some conflict resolution involved to avoid duplicates that took quite some time.

### CODE REVIEW

**Review by dragunoff — CHANGES_REQUESTED** [review]
I have some comments on the translation IDs. in #20207 I tried to establish a naming convention of *thing-id* where *thing* is usually the UI element (label, button, notification, dialog, dropdown, checkbox, etc.). Here the only prefixes are `text-`, `tooltiptext-` and `tooltipdesc-` and sometimes the *thing* is in the suffix (e.g. `text-new-map-bg-tileset-label`, `text-new-map-bg-cancel-button`).

The prefix `text-` is not a very useful one as it doesn't tell anything about the UI element. I suggest we try to match with the existing translations - `button-` for buttons, `label-` for non-interactive text. I'm on the fence about adding a `tooltip-` prefix because tooltips are often complex and consist of several labels.

Another I'd like to suggest is the use of attributes to group together related UI elements. This is something that was also established in #20207 (`dialog-`, `dropdown-`, `checkbox-`, `menu-`). Looking through the ftl file I see many places where this would be appropriate. A few examples:
```fluent
dialog-kick-client =
    .texta = You may also apply a temporary ban, preventing
    .textb = them from joining for the remainder of this game.
    .prevent-rejoining-checkbox = Temporarily Ban
    .confirm = Kick
    .cancel = Cancel
```

```fluent
button-command-bar-attack-move =
    .label = Attack Move
    .description = Selected units will move to the desired location
    and attack any enemies they encounter en route.
    Hold <(Ctrl)> while targeting to order an Assault Move
    that attacks any units or structures encountered en route.
    Left-click icon then right-click on target location.
```

```fluent
menu-production-type =
    .building = Buildings
    .defence = Support
    .infantry = Infantry
    .vehicle = Vehicles
    .aircraft = Aircraft
```

Thanks for continuing your work on localization - it's a monumental task :+1:

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameCashCounterLogic.cs [hunk 1]
```diff
@@ -28,7 +28,6 @@ public class IngameCashCounterLogic : ChromeLogic
 		readonly PlayerResources playerResources;
 		readonly LabelWithTooltipWidget cashLabel;
 		readonly CachedTransform<(int Resources, int Capacity), string> siloUsageTooltipCache;
-		readonly string cashTemplate;
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameCashCounterLogic.cs]: Shouldn't we keep formatting?

File: mods/common/chrome/connection.yaml [hunk 2]
```diff
@@ -18,22 +18,21 @@ Background@CONNECTIONFAILED_PANEL:
 			Y: 46
 			Width: PARENT_RIGHT
 			Height: 25
-			Text: Could not connect to AAA.BBB.CCC.DDD:EEEE
```
Comment (PunkPun) [inline mods/common/chrome/connection.yaml]: removed in the second commit so automatic extraction didn't detect this

File: mods/common/chrome/connection.yaml [hunk 3]
```diff
@@ -70,22 +69,21 @@ Background@CONNECTING_PANEL:
...
-			Text: Connecting
+			Text: label-connecting-panel-connecting-title
 			Align: Center
 			Font: Bold
 		Label@CONNECTING_DESC:
 			X: 0
 			Y: 61
 			Width: PARENT_RIGHT
 			Height: 25
-			Text: Connecting to AAA.BBB.CCC.DDD:EEEE...
```
Comment (PunkPun) [inline mods/common/chrome/connection.yaml]: same here

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+			var modData = Game.ModData = utility.ModData;
+
+			var types = modData.ObjectCreator.GetTypes();
+			var translatableFields = types.SelectMany(t => t.GetFields())
+				.Where(f => f.HasAttribute<TranslationReferenceAttribute>()).Distinct();
+
+			foreach (var chromeLayout in modData.Manifest.ChromeLayout)
+			{
+				modData.ModFiles.TryGetPackageContaining(chromeLayout, out var chromePackage, out var chromeName);
+				var chromePath = chromePackage.Name + "/" + chromeName;
```
Comment (PunkPun) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]:
```suggestion
				var chromePath = chromePackage.Name + Path.DirectorySeparatorChar + chromeName;
```

For windows compatibility?

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+				var fluent = chromeLayout.Split('|').First() + "|languages/chrome/en.ftl";
+				modData.ModFiles.TryGetPackageContaining(fluent, out var fluentPackage, out var fluentName);
+				var fluentPath = fluentPackage.Name + "/" + fluentName;
```
Comment (PunkPun) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]:
```suggestion
				var fluentPath = fluentPackage.Name + Path.DirectorySeparatorChar + fluentName;
```

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+			var types = modData.ObjectCreator.GetTypes();
+			var translatableFields = types.SelectMany(t => t.GetFields())
+				.Where(f => f.HasAttribute<TranslationReferenceAttribute>()).Distinct();
+
+			foreach (var chromeLayout in modData.Manifest.ChromeLayout)
+			{
+				modData.ModFiles.TryGetPackageContaining(chromeLayout, out var chromePackage, out var chromeName);
+				var chromePath = chromePackage.Name + "/" + chromeName;
+
+				var fluent = chromeLayout.Split('|').First() + "|languages/chrome/en.ftl";
```
Comment (PunkPun) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]: and here?

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+				using (var fluentWriter = new StreamWriter(fluentPath, append: true))
+				{
+					fluentWriter.WriteLine($"## {chromeLayout}");
+					var yaml = MiniYaml.FromFile(chromePath, false);
+
+					FromChromeLayout(ref yaml, null, translatableFields.Select(t => t.Name).Distinct(), null, fluentWriter);
+
+					using (var chromeLayoutWriter = new StreamWriter(chromePath))
+						chromeLayoutWriter.Write(yaml.WriteToString());
```
Comment (PunkPun) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]: perhaps write a newline for more consistent formating

**Review by PunkPun — COMMENTED** [review]
`make test` spits out tons of warnings. I suppose GH actions don't fail on warning

**Comment by PunkPun** [comment]
I recon for linting we'll probably have to do a similar structure to traits, splitting widgets into info and the main body.

Perhaps it doesn't need to be this drastic though, only translations separate?

File: mods/common/chrome/connection.yaml (hunk 2 above)
Comment (Mailaender) [reply mods/common/chrome/connection.yaml]: This is a misunderstanding on your side. It is already translated. https://github.com/OpenRA/OpenRA/blob/939f715e3c1cd82a080a26d57c29b5f049f6f4bb/OpenRA.Mods.Common/Widgets/Logic/ConnectionLogic.cs#L69

File: mods/common/chrome/connection.yaml (hunk 3 above)
Comment (Mailaender) [reply mods/common/chrome/connection.yaml]: You are wrong here as well.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameCashCounterLogic.cs (hunk 1 above)
Comment (Mailaender) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameCashCounterLogic.cs]: There is none:
https://github.com/OpenRA/OpenRA/blob/939f715e3c1cd82a080a26d57c29b5f049f6f4bb/mods/cnc/chrome/ingame.yaml#L1740

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs (hunk 6 above)
Comment (Mailaender) [reply OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]: No, this is OpenRA MiniYAML syntax.

**Review by PunkPun — COMMENTED** [review]
`make test` reports lots of unused translations in TD, and a few in D2K

**Comment by PunkPun** [comment]
And RA is complaining in some maps with
> Error: Missing Type: GiveBaseBuilderCrateActionInfo

It doesn't on bleed

**Comment by Mailaender** [comment]
Looks like you forgot to recompile as you are describing an unrelated recent change from #20728.

**Comment by PunkPun** [comment]
doesn't `make test` compile?

**Comment by PunkPun** [comment]
job logs also show errors for cnc and d2k

**Comment by Mailaender** [comment]
Well the CI does not spot your problem so it is clearly a cache/you problem.

**Comment by PunkPun** [comment]
rerunning tests again, this time RA did not have any issues

**Comment by Mailaender** [comment]
Sorry, the manual step is quite error-prone and has to be repeated for every requested refactor.

**Comment by PunkPun** [comment]
Could `Performance improvements.` changelog fit?

File: OpenRA.Mods.Common/Widgets/ButtonWidget.cs [hunk 8]
```diff
@@ -74,16 +77,16 @@ public ButtonWidget(ModData modData)
...
 			GetColorDisabled = () => TextColorDisabled;
 			GetContrastColorDark = () => ContrastColorDark;
 			GetContrastColorLight = () => ContrastColorLight;
 			OnMouseUp = _ => OnClick();
 			OnKeyPress = _ => OnClick();
 			IsHighlighted = () => Highlighted;
-			GetTooltipText = () => TooltipText;
-			GetTooltipDesc = () => TooltipDesc;
+			GetTooltipText = () => modData.Translation.GetString(TooltipText);
+			GetTooltipDesc = () => modData.Translation.GetString(TooltipDesc);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Widgets/ButtonWidget.cs]: We should definitely cache these. Translating all buttons every frame seems non-ideal

File: OpenRA.Mods.Common/Widgets/LabelWidget.cs [hunk 9]
```diff
@@ -37,9 +38,10 @@ public class LabelWidget : Widget
 		public Func<Color> GetContrastColorDark;
 		public Func<Color> GetContrastColorLight;
 
-		public LabelWidget()
+		[ObjectCreator.UseCtor]
+		public LabelWidget(ModData modData)
 		{
-			GetText = () => Text;
+			GetText = () => modData.Translation.GetString(Text);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Widgets/LabelWidget.cs]: Same here

**Review by PunkPun — COMMENTED** [review]
I suggest moving the 1st and 2nd commits into a separate PR, as well as part of the third that has changes to `LabelWithHighlightWidget`.

IMO this PR is still a WIP and not mergeable until we sort out linting. Considering this, I would suggest not bothering to rebase when widget changes happen. When I have more time I'll see what solutions I can come up to address liting.

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+			var modData = Game.ModData = utility.ModData;
+
+			var types = modData.ObjectCreator.GetTypes();
+			var translatableFields = types.SelectMany(t => t.GetFields())
+				.Where(f => f.HasAttribute<TranslationReferenceAttribute>()).Distinct();
+
+			foreach (var chromeLayout in modData.Manifest.ChromeLayout)
+			{
+				modData.ModFiles.TryGetPackageContaining(chromeLayout, out var chromePackage, out var chromeName);
+				var chromePath = chromePackage.Name + Path.DirectorySeparatorChar + chromeName;
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]:
```suggestion
				var chromePath = Path.Combine(chromePackage.Name, chromeName);
```
?

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+				var fluent = chromeLayout.Split('|').First() + "|languages/chrome/en.ftl";
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]: `string.Split` should return an array, so
```suggestion
				var fluent = chromeLayout.Split('|')[0] + "|languages/chrome/en.ftl";
```

Also, this is really weird as I assume that file does not exist yet. Is there no other way to get the fluent package? E.g. directly calling `OpenPackage`?

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+				modData.ModFiles.TryGetPackageContaining(fluent, out var fluentPackage, out var fluentName);
+				var fluentPath = fluentPackage.Name + Path.DirectorySeparatorChar + fluentName;
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]: `Path.Combine` here as well.

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+			foreach (var chromeLayout in modData.Manifest.ChromeLayout)
+			{
+				modData.ModFiles.TryGetPackageContaining(chromeLayout, out var chromePackage, out var chromeName);
+				var chromePath = Path.Combine(chromePackage.Name, chromeName);
+
+				var fluentPath = Path.Combine(Platform.EngineDir, "mods", modData.Manifest.Id, "languages/chrome/");
+				var fluentFile = Path.Combine(new Folder(fluentPath).Name, "en.ftl");
+				Directory.CreateDirectory(Path.GetDirectoryName(fluentPath));
+
+				using (var fluentWriter = new StreamWriter(fluentFile, append: true))
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]:
```suggestion
				// The "languages" folder should exist in every mod, so use it to determine the absolute path to place our new file in
				var fluent = modData.Manifest.Id + "|languages";
				var fluentPackage = modData.ModFiles.OpenPackage(fluent);
				var fluentPath = Path.Combine(fluentPackage.Name, "chrome/en.ftl");
				Directory.CreateDirectory(Path.GetDirectoryName(fluentPath));

				using (var fluentWriter = new StreamWriter(fluentPath, append: true))
```

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+					FromChromeLayout(yamlBuilder, null, translatableFields.Select(t => t.Name).Distinct(), null, fluentWriter);
+
+					using (var chromeLayoutWriter = new StreamWriter(chromePath))
+						chromeLayoutWriter.WriteLine(yamlBuilder.WriteToString());
+				}
+			}
+		}
+
+#pragma warning disable CA1851 // Possible multiple enumerations of 'IEnumerable' collection
+		internal static void FromChromeLayout(List<MiniYamlNodeBuilder> nodes, MiniYamlNodeBuilder parent, IEnumerable<string> translatables, string container, StreamWriter fluentFile)
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs]:
```suggestion
		internal static void FromChromeLayout(List<MiniYamlNodeBuilder> nodes, MiniYamlNodeBuilder parent, string[] translatables, string container, StreamWriter fluentFile)
```

and then call `.ToArray()` in line 53 to silence the warning.

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+		{
+			var parentNode = parent != null && parent.Key != null ? parent.Key.Split('@') : null;
+			var parentType = parent != null && parent.Key != null ? parentNode.First() : null;
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs:64]:
```suggestion
			var parentType = parent != null && parent.Key != null && parentNode.Length > 0 ? parentNode[0] : null;
```

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+			var parentLabel = parent != null && parent.Key != null ? parentNode.Last() : null;
```
Comment (abcdefg30) [inline OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs:65]:
```suggestion
			var parentLabel = parent != null && parent.Key != null && parentNode.Length > 0 ? parentNode[^1] : null;
```

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs (hunk 15 above)
Comment (Mailaender) [reply OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs:64]: I think this is less readable.

Comment (abcdefg30) [reply OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs:64]: Maybe so, but its LINQ and slower than direct access.

Comment (PunkPun) [reply OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs:64]: tbf perf doesn't matter here at all

**Review by RoosterDragon — APPROVED** [review]
Good idea with the automation. Checked various UI in all 4 mods and no obvious issues with missing text or the like. Translations files and placeholders in yaml all seem to have reasonable naming to me. No code issues I would suggest changes on.

File: OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs (hunk 15 above)
Comment (Mailaender) [reply OpenRA.Game/UtilityCommands/ExtractChromeStrings.cs:64]: It is more consistent with the other lines this way.

File: OpenRA.Mods.Common/Lint/CheckTranslationReference.cs [hunk 17]
```diff
@@ -135,6 +136,13 @@ void ILintPass.Run(Action<string> emitError, Action<string> emitWarning, ModData
 				}
 			}
 
+			var chromeFields = new[] { "Text", "TooltipText", "TooltipDesc" };
+			foreach (var filename in modData.Manifest.ChromeLayout)
+			{
+				var nodes = MiniYaml.FromStream(modData.DefaultFileSystem.Open(filename));
+				CheckChrome(nodes, chromeFields, translation, language, emitError);
+			}
+
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Lint/CheckTranslationReference.cs:145]: I now see the discussion was solely contained in discord. I have mentioned it multiple times that this should be using reflection instead, I don't see next release without the proper linting. A good example of how to do this is the chrome hotkey linting

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Lint/CheckTranslationReference.cs:145]: Is it worth noting that as a follow-up in the tracking issue #10475? Getting this large batch of translations in to reduce the rebase toil would be a win.

Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckTranslationReference.cs:145]: I'm waiting on #21003 to be merged first as I want less conflicts between prep and bleed

**Comment by PunkPun** [comment]
needs a rebase

**Pull request closed without merging**